test/tlsconnection.js

```diff
@@ -0,0 +1,1160 @@
+      await client.recv(serverHello);
+    }, TLSError, 'ILLEGAL_PARAMETER');
+  });
+
+  it('errors if the ServerHello contained unexpected extensions', async () => {
+    const serverHello = testHelpers.makePlaintextRecord({
+      content: await testHelpers.makeServerHelloMessage({
+        extensions: [
+          testHelpers.makeSupportedVersionsExtension(0x0304),
+          testHelpers.makePskKeyExchangeModesExtension([]),
+          testHelpers.makePreSharedKeyExtension(0)
+        ]
+      }),
+      type: 22
+    });
+    await assert.throwsAsync(async () => {
+      await client.recv(serverHello);
+    }, TLSError, 'UNSUPPORTED_EXTENSION');
+  });
+
+  it('errors if the ServerHello contained duplicate extensions', async () => {
+    const serverHello = testHelpers.makePlaintextRecord({
+      content: await testHelpers.makeServerHelloMessage({
+        extensions: [
+          testHelpers.makeSupportedVersionsExtension(0x0304),
+          testHelpers.makeSupportedVersionsExtension(0x0305),
+          testHelpers.makePreSharedKeyExtension(0)
+        ]
+      }),
+      type: 22
+    });
+    await assert.throwsAsync(async () => {
+      await client.recv(serverHello);
+    }, TLSError, 'DECODE_ERROR');
+  });
+
+  it('errors if ServerHello and EncryptedExtensions appear in the same record', async () => {
+    const contentBuf = new BufferWriter();
+    contentBuf.writeBytes(await testHelpers.makeServerHelloMessage({
+      extensions: [
+        testHelpers.makeSupportedVersionsExtension(0x0304),
+        testHelpers.makePreSharedKeyExtension(0)
+      ]
+    }));
+    contentBuf.writeBytes(new EncryptedExtensions().toBytes());
+    const serverHello = testHelpers.makePlaintextRecord({
+      content: contentBuf.flush(),
+      type: 22
+    });
+    await assert.throwsAsync(async () => {
+      await client.recv(serverHello);
+    }, TLSError, 'UNEXPECTED_MESSAGE');
+  });
+});
+
+
+describe('the handshake between ClientConnection and ServerConnection', () => {
+  it('works without any mocking', async () => {
+    const CLIENT_SENT = [], SERVER_SENT = [];
+    const server = await ServerConnection.create(TEST_VECTORS.PSK, TEST_VECTORS.PSK_ID, data => SERVER_SENT.push(data));
+    const client = await ClientConnection.create(TEST_VECTORS.PSK, TEST_VECTORS.PSK_ID, data => CLIENT_SENT.push(data));
+    assert.equal(CLIENT_SENT.length, 1);
+    await server.recv(CLIENT_SENT[0]);
+    assert.equal(SERVER_SENT.length, 3);
+    await client.recv(SERVER_SENT[0]);
+    await client.recv(SERVER_SENT[1]);
+    await client.recv(SERVER_SENT[2]);
+    assert.equal(CLIENT_SENT.length, 2);
+    await server.recv(CLIENT_SENT[1]);
+    assert.ok(bytesAreEqual(
+      server._keyschedule.clientApplicationTrafficSecret,
+      client._keyschedule.clientApplicationTrafficSecret
+    ));
+    assert.ok(bytesAreEqual(
+      server._keyschedule.serverApplicationTrafficSecret,
+      client._keyschedule.serverApplicationTrafficSecret
+    ));
+  });
```

Comment:
Can you add tests here for the promise I talked about.

Also, can you add a test that includes calls to send().  It would be good if you could also test that you can close in one direction then continue sending and receiving in the other direction.